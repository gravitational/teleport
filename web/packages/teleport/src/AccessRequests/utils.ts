/**
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import {
  addDays,
  addHours,
  Duration,
  intervalToDuration,
  isAfter,
  isBefore,
} from 'date-fns';

type TimeDuration = {
  timestamp: number;
  duration: Duration;
};

// Round the duration to the nearest 10 minutes
// Example:
// 9m -> 10m
// 10m -> 10m
// 11m -> 10m
// 15m -> 20m
// 1d -> 1d
// 1d 1h -> 1d 1h
// The only exception is 0m, which is rounded to 10m
export function roundToNearestTenMinutes(date: Duration): Duration {
  let minutes = date.minutes;
  let roundedMinutes = Math.round(minutes / 10) * 10; // Round to the nearest 10
  if (roundedMinutes === 0 && !date.days && !date.hours) {
    // Do not round down to 0. This
    roundedMinutes = 10;
  }

  // 60 minutes == 1 hour
  // Prevent displaying time as eg: `5 hrs and 60 mins`
  // to `6 hrs`.
  if (roundedMinutes === 60) {
    if (!date.hours) {
      date.hours = 0;
    }
    date.hours += 1;
    roundedMinutes = 0;
  }
  date.minutes = roundedMinutes;
  date.seconds = 0;

  return date;
}

// Generate a list of middle values between start and end. The first value is the
// session TTL that is rounded to the nearest hour. The rest of the values are
// rounded to the nearest day. Example:
//
// created: 2021-09-01T00:00:00.000Z
// start: 2021-09-01T01:00:00.000Z
// end: 2021-09-03T00:00:00.000Z
// now: 2021-09-01T00:00:00.000Z
//
// returns: [1h, 1d, 2d, 3d]
export function middleValues(
  created: Date,
  start: Date,
  end: Date
): TimeDuration[] {
  const getInterval = (d: Date) =>
    roundToNearestTenMinutes(
      intervalToDuration({
        start: created,
        end: d,
      })
    );

  const points: Date[] = [start];

  if (isAfter(addDays(created, 1), end)) {
    // Add all possible options to the list. This covers the case when the
    // max duration is less than 24 hours.
    if (isBefore(addHours(points[points.length - 1], 1), end)) {
      points.push(end);
    }

    return points.map(d => ({
      timestamp: d.getTime(),
      duration: getInterval(d),
    }));
  }

  points.push(addDays(created, 1));

  // I also prefer while(true), but our linter doesn't
  for (;;) {
    const next = addHours(points[points.length - 1], 24);
    // Allow next == end
    if (next > end) {
      break;
    }
    points.push(next);
  }

  return points.map(d => ({
    timestamp: d.getTime(),
    duration: getInterval(d),
  }));
}

// Generate a list of middle values between now and the request TTL.
export function requestTtlMiddleValues(
  created: Date,
  requestTtl: Date
): TimeDuration[] {
  const getInterval = (d: Date) =>
    roundToNearestTenMinutes(
      intervalToDuration({
        start: created,
        end: d,
      })
    );

  if (isAfter(addHours(created, 1), requestTtl)) {
    return [
      {
        timestamp: requestTtl.getTime(),
        duration: getInterval(requestTtl),
      },
    ];
  }

  const points: Date[] = [];
  // Staggered hour options, up to 1 week.
  const hourOptions = [1, 2, 3, 4, 6, 8, 12, 18, 24, 30, 168];

  for (const h of hourOptions) {
    const t = addHours(created, h);
    if (isAfter(t, requestTtl)) {
      break;
    }
    points.push(t);
  }

  return points.map(d => ({
    timestamp: d.getTime(),
    duration: getInterval(d),
  }));
}
