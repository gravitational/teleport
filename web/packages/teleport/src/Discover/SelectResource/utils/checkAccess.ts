/**
 * Teleport
 * Copyright (C) 2025  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { Acl } from 'teleport/services/user';

import { ResourceKind } from '../../Shared';
import { ResourceSpec } from '../types';

function checkHasAccess(acl: Acl, resourceKind: ResourceKind) {
  const basePerm = acl.tokens.create;
  if (!basePerm) {
    return false;
  }

  switch (resourceKind) {
    case ResourceKind.Application:
      return acl.appServers.read && acl.appServers.list;
    case ResourceKind.Database:
      return acl.dbServers.read && acl.dbServers.list;
    case ResourceKind.Desktop:
      return acl.desktops.read && acl.desktops.list;
    case ResourceKind.Kubernetes:
      return acl.kubeServers.read && acl.kubeServers.list;
    case ResourceKind.Server:
      return acl.nodes.list;
    case ResourceKind.SamlApplication:
      return acl.samlIdpServiceProvider.create;
    case ResourceKind.ConnectMyComputer:
      // This is probably already true since without this permission the user wouldn't be able to
      // add any other resource, but let's just leave it for completeness sake.
      return acl.tokens.create;
    default:
      return false;
  }
}

export function addHasAccessField(
  acl: Acl,
  resources: ResourceSpec[]
): ResourceSpec[] {
  return resources.map(r => {
    const hasAccess = checkHasAccess(acl, r.kind);
    switch (r.kind) {
      case ResourceKind.Database:
        return { ...r, dbMeta: { ...r.dbMeta }, hasAccess };
      default:
        return { ...r, hasAccess };
    }
  });
}
