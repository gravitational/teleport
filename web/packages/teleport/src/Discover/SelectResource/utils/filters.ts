/**
 * Teleport
 * Copyright (C) 2025  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { Platform } from 'design/platform';

import { ResourceKind } from 'teleport/Discover/Shared';
import { AuthType } from 'teleport/services/user';

import { SelectResourceSpec } from '../resources';

export function filterBySupportedPlatformsAndAuthTypes(
  platform: Platform,
  authType: AuthType,
  resources: SelectResourceSpec[]
) {
  return resources.filter(resource => {
    const resourceSupportsPlatform =
      !resource.supportedPlatforms?.length ||
      resource.supportedPlatforms.includes(platform);

    const resourceSupportsAuthType =
      !resource.supportedAuthTypes?.length ||
      resource.supportedAuthTypes.includes(authType);

    return resourceSupportsPlatform && resourceSupportsAuthType;
  });
}

export const resourceTypeOptions = [
  { value: 'app', label: 'Applications' },
  { value: 'db', label: 'Database' },
  { value: 'desktops', label: 'Desktops' },
  { value: 'kube', label: 'Kubernetes' },
  { value: 'server', label: 'SSH' },
] as const satisfies { value: string; label: string }[];

type ResourceType = Extract<
  (typeof resourceTypeOptions)[number],
  { value: string }
>['value'];

export const hostingPlatformOptions = [
  { value: 'aws', label: 'Amazon Web Services (AWS)' },
  { value: 'azure', label: 'Microsoft Azure' },
  { value: 'gcp', label: 'Google Cloud Services (GCP)' },
  { value: 'self-hosted', label: 'Self-Hosted' },
] as const satisfies { value: string; label: string }[];

type HostingPlatform = Extract<
  (typeof hostingPlatformOptions)[number],
  { value: string }
>['value'];

export type Filters = {
  resourceTypes?: ResourceType[];
  hostingPlatforms?: HostingPlatform[];
};

export function filterResources(
  resources: SelectResourceSpec[],
  filters: Filters
) {
  if (
    !resources.length &&
    !filters.resourceTypes &&
    !filters.hostingPlatforms
  ) {
    return resources;
  }

  let filtered = [...resources];
  if (filters.resourceTypes.length) {
    const resourceTypes = filters.resourceTypes;
    filtered = filtered.filter(r => {
      if (
        resourceTypes.includes('app') &&
        (r.kind === ResourceKind.Application ||
          r.kind === ResourceKind.SamlApplication)
      ) {
        return true;
      }
      if (resourceTypes.includes('db') && r.kind === ResourceKind.Database) {
        return true;
      }
      if (
        resourceTypes.includes('desktops') &&
        r.kind === ResourceKind.Desktop
      ) {
        return true;
      }
      if (
        resourceTypes.includes('kube') &&
        r.kind === ResourceKind.Kubernetes
      ) {
        return true;
      }
      if (
        resourceTypes.includes('server') &&
        (r.kind === ResourceKind.Server ||
          r.kind === ResourceKind.ConnectMyComputer)
      ) {
        return true;
      }
    });
  }

  if (filters.hostingPlatforms.length) {
    const hostingPlatforms = filters.hostingPlatforms;
    filtered = filtered.filter(r => {
      if (
        hostingPlatforms.includes('aws') &&
        r.keywords.some(k => k.toLowerCase().includes('aws'))
      ) {
        return true;
      }
      if (
        hostingPlatforms.includes('azure') &&
        r.keywords.some(k => k.toLowerCase().includes('azure'))
      ) {
        return true;
      }
      if (
        hostingPlatforms.includes('gcp') &&
        r.keywords.some(k => k.toLowerCase().includes('gcp'))
      ) {
        return true;
      }
      if (
        hostingPlatforms.includes('self-hosted') &&
        r.keywords.some(k => k.toLowerCase().includes('self-hosted'))
      ) {
        return true;
      }
    });
  }

  return filtered;
}
