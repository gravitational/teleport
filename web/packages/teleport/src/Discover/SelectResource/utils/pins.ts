/**
 * Teleport
 * Copyright (C) 2025  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { UserPreferences } from 'gen-proto-ts/teleport/userpreferences/v1/userpreferences_pb';

import { DiscoverGuideId } from 'teleport/services/userPreferences/discoverPreference';

import { SelectResourceSpec } from '../resources';

export const defaultPins = [
  DiscoverGuideId.ConnectMyComputer,
  DiscoverGuideId.Kubernetes,
  DiscoverGuideId.ApplicationWebHttpProxy,
  // TODO(kimlisa): add linux server, once they are all consolidated into one
];

/**
 * Only returns the defaults of resources that is available to the user.
 */
export function getDefaultPins(availableResources: SelectResourceSpec[]) {
  return availableResources
    .filter(r => defaultPins.includes(r.id))
    .map(r => r.id);
}

/**
 * Returns pins from preference if any or default pins.
 */
export function getPins(preferences: DiscoverResourcePreference) {
  if (!preferences.discoverResourcePreferences) {
    return [];
  }

  if (!preferences.discoverResourcePreferences.discoverGuide) {
    return defaultPins;
  }

  return preferences.discoverResourcePreferences.discoverGuide.pinned || [];
}

export type DiscoverResourcePreference = Partial<
  Pick<UserPreferences, 'discoverResourcePreferences'>
>;
