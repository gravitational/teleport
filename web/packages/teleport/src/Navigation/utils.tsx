/**
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import {
  Icon,
  NavigationItemSize,
  SmallIcon,
} from 'teleport/Navigation/common';

import type { TeleportFeature } from 'teleport/types';

export function getIcon(feature: TeleportFeature, size: NavigationItemSize) {
  switch (size) {
    case NavigationItemSize.Large:
      return <Icon>{<feature.navigationItem.icon />}</Icon>;

    case NavigationItemSize.Small:
      return <SmallIcon>{<feature.navigationItem.icon />}</SmallIcon>;
  }
}
