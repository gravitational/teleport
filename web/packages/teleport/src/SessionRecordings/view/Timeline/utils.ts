/**
 * Teleport
 * Copyright (C) 2025  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import type { SessionRecordingThumbnail } from 'teleport/services/recordings';

const zoomLevelWithCursor = 2;
const zoomLevelWithoutCursor = 1;

// calculateThumbnailViewport calculates the portion of the thumbnail image to display
// based on the cursor position and whether the cursor is visible.
// It returns the source coordinates and dimensions to be used in drawing the image,
// as well as the zoom level applied.
export function calculateThumbnailViewport(
  thumbnail: SessionRecordingThumbnail,
  width: number,
  height: number
) {
  // Use different zoom levels based on cursor visibility
  const zoomLevel = thumbnail.cursorVisible
    ? zoomLevelWithCursor
    : zoomLevelWithoutCursor;

  const visibleWidthPercent = 100 / zoomLevel;
  const visibleHeightPercent = 100 / zoomLevel;

  let bgPosX: number;
  let bgPosY: number;

  if (thumbnail.cursorVisible) {
    // Calculate cursor position as percentage
    const cursorPercentX = (thumbnail.cursorX / thumbnail.cols) * 100;
    const cursorPercentY = (thumbnail.cursorY / thumbnail.rows) * 100;

    // Calculate the top-left position percentage to center on cursor
    bgPosX = Math.max(
      0,
      Math.min(
        100 - visibleWidthPercent,
        cursorPercentX - visibleWidthPercent / 2
      )
    );
    bgPosY = Math.max(
      0,
      Math.min(
        100 - visibleHeightPercent,
        cursorPercentY - visibleHeightPercent / 2
      )
    );
  } else {
    // Center the viewport when cursor is not visible
    bgPosX = (100 - visibleWidthPercent) / 2;
    bgPosY = (100 - visibleHeightPercent) / 2;
  }

  // Calculate source dimensions
  const sourceWidth = width / zoomLevel;
  const sourceHeight = height / zoomLevel;

  // Convert percentages to pixel coordinates on the source image
  const maxSourceX = width - sourceWidth;
  const maxSourceY = height - sourceHeight;

  const sourceX = (bgPosX / (100 - visibleWidthPercent)) * maxSourceX || 0;
  const sourceY = (bgPosY / (100 - visibleHeightPercent)) * maxSourceY || 0;

  return {
    sourceX,
    sourceY,
    sourceWidth,
    sourceHeight,
    zoomLevel,
  };
}
