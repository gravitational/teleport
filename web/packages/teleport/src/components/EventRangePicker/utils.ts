/**
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { endOfDay, startOfDay, subDays } from 'date-fns';

export function getRangeOptions(): EventRange[] {
  return [
    {
      name: 'Today',
      from: startOfDay(new Date()),
      to: endOfDay(new Date()),
    },
    {
      name: '7 days',
      from: startOfDay(subDays(new Date(), 6)),
      to: endOfDay(new Date()),
    },
    {
      name: 'Custom Range...',
      isCustom: true,
      from: new Date(),
      to: new Date(),
    },
  ];
}

export type EventRange = {
  from: Date;
  to: Date;
  isCustom?: boolean;
  name?: string;
};
