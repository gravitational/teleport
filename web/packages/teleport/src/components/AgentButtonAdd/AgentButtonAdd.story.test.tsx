/**
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react';

import { render } from 'design/utils/testing';

import {
  CanCreate,
  CannotCreate,
  CannotCreateVowel,
  OnLeaf,
  OnLeafVowel,
} from './AgentButtonAdd.story';

test('can create', () => {
  const { container } = render(<CanCreate />);
  expect(container).toMatchSnapshot();
});

test('cannot create', () => {
  const { container } = render(<CannotCreate />);
  expect(container).toMatchSnapshot();
});

test('cannot create when resource starts with vowel', () => {
  const { container } = render(<CannotCreateVowel />);
  expect(container).toMatchSnapshot();
});

test('on leaf cluster', () => {
  const { container } = render(<OnLeaf />);
  expect(container).toMatchSnapshot();
});

test('on leaf cluster when resource starts with vowel', () => {
  const { container } = render(<OnLeafVowel />);
  expect(container).toMatchSnapshot();
});
