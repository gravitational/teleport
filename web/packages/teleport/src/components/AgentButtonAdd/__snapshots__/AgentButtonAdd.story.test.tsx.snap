// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create 1`] = `
.c0 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  color: #000000;
  background: #9F85FF;
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  width: 240px;
  text-transform: none;
}

.c0:hover,
.c0:focus {
  background: #B29DFF;
}

.c0:active {
  background: #C5B6FF;
}

.c0:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
  cursor: auto;
}

<div>
  <a
    href="/web/discover"
    style="text-decoration: none;"
  >
    <button
      class="c0"
      kind="primary"
      title=""
      width="240px"
    >
      Add server
    </button>
  </a>
</div>
`;

exports[`cannot create 1`] = `
.c0 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  color: #000000;
  background: #9F85FF;
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  width: 240px;
  text-transform: none;
}

.c0:hover,
.c0:focus {
  background: #B29DFF;
}

.c0:active {
  background: #C5B6FF;
}

.c0:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
  cursor: auto;
}

<div>
  <a
    href="/web/discover"
    style="text-decoration: none;"
  >
    <button
      class="c0"
      disabled=""
      kind="primary"
      title="You do not have access to add a server"
      width="240px"
    >
      Add server
    </button>
  </a>
</div>
`;

exports[`cannot create when resource starts with vowel 1`] = `
.c0 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  color: #000000;
  background: #9F85FF;
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  width: 240px;
  text-transform: none;
}

.c0:hover,
.c0:focus {
  background: #B29DFF;
}

.c0:active {
  background: #C5B6FF;
}

.c0:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
  cursor: auto;
}

<div>
  <a
    href="/web/discover"
    style="text-decoration: none;"
  >
    <button
      class="c0"
      disabled=""
      kind="primary"
      title="You do not have access to add an application"
      width="240px"
    >
      Add application
    </button>
  </a>
</div>
`;

exports[`on leaf cluster 1`] = `
.c0 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  color: #000000;
  background: #9F85FF;
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  width: 240px;
  text-transform: none;
}

.c0:hover,
.c0:focus {
  background: #B29DFF;
}

.c0:active {
  background: #C5B6FF;
}

.c0:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
  cursor: auto;
}

<div>
  <a
    href="/web/discover"
    style="text-decoration: none;"
  >
    <button
      class="c0"
      disabled=""
      kind="primary"
      title="Adding a server to a leaf cluster is not supported"
      width="240px"
    >
      Add server
    </button>
  </a>
</div>
`;

exports[`on leaf cluster when resource starts with vowel 1`] = `
.c0 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  color: #000000;
  background: #9F85FF;
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  width: 240px;
  text-transform: none;
}

.c0:hover,
.c0:focus {
  background: #B29DFF;
}

.c0:active {
  background: #C5B6FF;
}

.c0:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
  cursor: auto;
}

<div>
  <a
    href="/web/discover"
    style="text-decoration: none;"
  >
    <button
      class="c0"
      disabled=""
      kind="primary"
      title="Adding an server to a leaf cluster is not supported"
      width="240px"
    >
      Add server
    </button>
  </a>
</div>
`;
